### PR TITLE
Better dependency on npm-package dex-contracts

### PR DESCRIPTION
**Pull request by josojo**
This PR depends on the other [PR here:](https://github.com/gnosis/dex-contracts/pull/927)

It imports the batchexchange dependency as suggested by the other PR

Testplan: 
See other PR.

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -48,6 +48,7 @@
     "@gnosis.pm/util-contracts": "^2.0.6",
     "@openzeppelin/contracts": "=2.5.1",
     "@truffle/contract": "4.2.15",
+    "solidity-bytes-utils": "0.0.9",
```
Comment (josojo) [inline package.json:51]: >"@gnosis.pm/dex-contracts": "^0.4.1",

this will need to be changed to 
"@gnosis.pm/dex-contracts": "^0.4.2", once the dex-contracts package is published

File: README.md [hunk 2]
```diff
@@ -21,6 +21,7 @@ Build Truffle artifacts:
 
 ```sh
 yarn build
+yarn copy-artifacts
```
[the remark this answers is not in the document]
Comment (fedgiac) [reply README.md:24]: Also, while I think it currently shouldn't matter because the artifacts from `dex-contracts` should have the various network addresses injected already(?), wouldn't it be better to copy the artifacts before `networks-inject` (which is part of `build`)?

[the remark this answers is not in the document]
Comment (josojo) [reply README.md:24]: I thought about making it part of the build process. But there is an issue; imagine you have already deployed it against ganache and then you change the FleetDeployment contract and hence want to rebuild it. This would erase your current migration with the addresses, as this copy command would overwrite your current artifact.

But you are probably saying that this is an edge case that is very unlikely to happen as a contract change will most likely always require a complete immigration of the contracts. Hm.

File: migrations/2_dev_dependencies.js [hunk 3]
```diff
@@ -4,14 +4,20 @@ const ProxyFactory = artifacts.require("./GnosisSafeProxyFactory.sol")
...
+    deployer,
+    network,
+    account: accounts[0],
+    web3,
+  })
+
+  const Artifactor = require("@truffle/artifactor")
+  const artifactor = new Artifactor("build/contracts/")
+  await artifactor.save(artefact)
+
```
[the remark this answers is not in the document]
Comment (josojo) [reply migrations/2_dev_dependencies.js:19]: If you run this migration against mainnet, you would not migrate a new contract, but you would just get the original BatchExchange from the npm package, because of this line: https://github.com/gnosis/dex-contracts/blob/master/src/migration/PoC_dfusion.js#L46

Hence, I don't see the fundamental risk and I think that it is nice to have it automatically copied, and we can't forget to do the copy statement.

File: package.json [hunk 4]
```diff
@@ -20,16 +20,18 @@
...
     "build": "yarn compile && yarn networks-inject",
+    "copy-artifacts": "cp ./node_modules/@gnosis.pm/dex-contracts/build/contracts/BatchExchange.json ./build/contracts",
     "compile": "truffle compile",
     "testnet": "ganache-cli --gasLimit 8e6",
     "migrate": "truffle migrate",
     "verify": "truffle run verify",
     "flatten": "export FLATTENED_DIR=\"./build/flattenedContracts\" && mkdir -p $FLATTENED_DIR && npx sol-merger \"contracts/*.sol\" $FLATTENED_DIR && echo \"Flattened contracts stored in $FLATTENED_DIR\" && unset FLATTENED_DIR",
-    "networks-inject": "CONF_FILE=$(pwd)'/migration_conf.js'  node node_modules/@gnosis.pm/util-contracts/src/inject_network_info.js",
-    "networks-extract": "CONF_FILE=$(pwd)'/migration_conf.js'  node node_modules/@gnosis.pm/util-contracts/src/extract_network_info.js"
+    "networks-inject": "CONF_FILE=$(pwd)'/migrations/migration_conf.js'  node node_modules/@gnosis.pm/util-contracts/src/inject_network_info.js  && CONF_FILE=$(pwd)'/migrations/migration_conf_dex-contracts.js'  node node_modules/@gnosis.pm/util-contracts/src/inject_network_info.js",
```
[the remark this answers is not in the document]
Comment (josojo) [reply package.json:29]: I agree! But this requires rewriting parts of @gnosis.pm/util-contracts, which is out of scope for this PR. Or do you see an easy solution for that?

**Comment by josojo** [comment]
> I like that we don't rely on `Contract` anymore. Do we still to depend on `@truffle/contract`?

We still need have it for the dependency:    ` const WETH = Contract(require("canonical-weth/build/contracts/WETH9"))`

File: package.json (hunk 4 above)
[the remark this answers is not in the document]
Comment (fedgiac) [reply package.json:29]: Good point, I don't see how it can be done with just Javascript without touching the dependency directly. I tried to come up with another way to remove the redundant config in PR #419, but I'm not sure how much of an improvement it would be. Feel free to merge or reject as you prefer.

File: migrations/2_dev_dependencies.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (fedgiac) [reply migrations/2_dev_dependencies.js:19]: Ah, I didn't see that! Then it's good as it is.

File: README.md (hunk 2 above)
[the remark this answers is not in the document]
Comment (fedgiac) [reply README.md:24]: The reason I suggest to merge it with `yarn build` is that it's a useless step for most of our users, since they would possibly don't even know what Ganache is, while the very few people who would want to use Ganache probably know how to read a `package.json`.
In case you want to keep the two separate anyway, the tutorial needs to be updated!

File: package.json [hunk 5]
```diff
@@ -19,7 +19,8 @@
     "prettier": "prettier --write './**/*.js'",
     "pretty-check": "prettier --check './**/*.js'",
     "test": "truffle test",
-    "build": "yarn compile && yarn networks-inject",
+    "build": "yarn compile && yarn copy-artifacts && yarn networks-inject",
+    "copy-artifacts": "cp ./node_modules/@gnosis.pm/dex-contracts/build/contracts/BatchExchange.json ./build/contracts",
```
Comment (fleupold) [inline package.json:23]: Why is this needed? So that we don't need to run a migration step and still can access the artefact with the direct require?

Comment (josojo) [reply package.json:23]: Yes! People just wanting to setup up bracket strategies should not go through the migration step
